### ISSUE TITLE
Paket local does not work with new csproj format

### ISSUE THREAD

**Issue by OsirisTerje** [fsprojects/Paket#3116]
### Description

Given a project using the new csproj format.  When a paket.local file is enabled, pointing one of the dependencies to a local package, the new csproj still uses the non-local package.

### Repro steps

A repro repository is uploaded at https://github.com/KDISim/PaketNewCsProjIssue  

The repo consists of two solutions.  Use the solution under source/MyTests/MyTests,sln.  
The solution has two csproj projects, one using the old project format, and one using the new csproj format.  Each project as a single unit test, both these tests are identical.  
Compile this, and you see the two unit tests going green. 

Then rename the paket.local.txt to paket.local,   do a paket restore, and a full rebuild. 
The local package has a change that should cause both tests to go red. 
But now you should see that only the test in the old csproj project goes red, the one in the new csproj stays green - indicating it doesn't use the local package, but still uses the non-local. 

If I check the csproj files using the MSBuild Structured log viewer, it also looks like the new csproj format doesn't evaluate to use the local package.

**Comment by forki** [comment]
paket.local is not yet supported for new style projects and we don't have an idea yet how to solve it.

**Comment by OsirisTerje** [comment]
What is the underlying issue?   
Isn't this handled through the restore targets file?    Or are there some behaviour in the new csproj format that circumvents it?

**Comment by forki** [comment]
yes it's the new csproj format. they always draw the package from that global cache folder - and we can't just overwrite that because then other solutions on your PC would also get the modified stuff

**Comment by OsirisTerje** [comment]
I tried to move the local package into the global cache folder (under %userprofile%/.nuget), and also tried pointing the paket.local source to the same place there, but it still would not work the new format, so it seems there is more than just the global cache that is blocking this.

**Comment by matthid** [comment]
It's probably because we forward all references to NuGet and NuGet is doing something we don't like?

**Comment by mclark1129** [comment]
So then does that mean with the SDK style projects, the contents of the `packages/` folder is completely ignored by Paket?  Regardless of whether or not we're using any local overrides?

**Comment by AlexBatsev** [comment]
Hi. Is there any progress on this issue? I'm really interested in having it fixed ASAP.
Thank you.

**Comment by forki** [comment]
@AlexBatsev this is an dotnet sdk issue.

**Comment by reijerh** [comment]
@forki I can open a github issue in one of the .NET repos if you like, I'm guessing this goes in `dotnet/cli`? Or `NuGet`? Or... `dotnet/sdk`?

I'm not 100% sure what I should ask for though, I guess you need a way to specify in which folder the dotnet build tool should look for NuGet packages?

@OsirisTerje did you by any chance figure out a work-around or other workable solution?

**Comment by OsirisTerje** [comment]
@reijerh Unfortunately not.

**Comment by vdurante** [comment]
Almost 2 years later... is there any workaround for this issue?

**Comment by vdurante** [comment]
@forki why not simply deleting the package from global cache?

If anyone is interested, I have built a very simple Powershell script that works for my scenario. If anyone else is interested, here is how to use it:

The scenario: I want my peers to resolve the package from GitHub. I want me to resolve it from my local development folder. **I need the solution to be as fast as possible since I will be developing 2 different solutions at the same time**.

paket.dependencies
```
source https://api.nuget.org/v3/index.json

storage: none
framework: netcoreapp3.1, netstandard2.0, netstandard2.1

git https://github.com/vdurante/VDurante.MyLib.git master build: "build.cmd", Packages: /bin/, OS: windows

# ...
# all packages declared here
# ...

nuget VDurante.MyLib

group LOCAL
	source C:\repos\VDurante\VDurante.MyLib\bin
	nuget VDurante.MyLib
```

paket.references
```
# ...
# all packages declared here
# ...

group LOCAL
    VDurante.MyLib
```

paket-local.ps1 (same directory of paket.dependencies)
```
function Get-PaketLocals {
    $localScope = $false

    Foreach($line in Get-Content -Path "./paket.dependencies"){
        $line = $line.Trim()

        if($line.StartsWith("//") -or $line.StartsWith("#")){            
            continue;
        }

        if($line.StartsWith("group")){
            $line = $line.Substring(5, $line.Length -5).Trim();
            if($line.ToUpperInvariant() -eq "LOCAL"){
                $localScope = $true
            } else {
                $localScope = $false
            }
            continue;
        }

        if(!$localScope -or !$line.StartsWith("nuget")){
            continue;
        }

        $package = $line.Substring(5, $line.Length - 5).Trim();
        
        Write-Output $package
    }
}

function Clear-PackageCache($package){
    [String]$cachePath = dotnet nuget locals global-packages --list
    $cachePath = $cachePath.Substring(16, $cachePath.Length - 16).Trim().TrimEnd("/").TrimEnd("\");    

    if(Test-Path $cachePath\$package  -PathType Any){
        Remove-Item -Recurse -Force $cachePath\$package
    }
}

Foreach($package in Get-PaketLocals){
    Clear-PackageCache $package
}

paket restore -g LOCAL --force
```

Run `powershell -File paket-local.ps1` (Command Prompt) or `.\paket-local.ps1` (PowerShell) and it should work. It will fast reload the dependencies from `LOCAL` group.

I don't advise committing and pushing any of this, since it will crash your peer's local development environments. If you need a "better" solution, you could rewrite the powershell script above to do the following logic:

1. create a backup of `paket.dependencies`, all `paket.references` files and `paket.lock`
2. clear nuget's cache from `global-packages` (already implemented), `http-cache`, and `temp` (this could be necessary under certain conditions that I am unaware of)
2. implement your own flavor of `paket.local`, such as `paket.core.local` with same syntax as `paket.local`
3. read from this file and write LOCAL groups into both `paket.dependencies` and `paket.references`
4. run `packet install` and `paket restore -g LOCAL --force` commands
5. restore `paket.dependencies`, all `paket.references` files and `paket.lock`

Good luck

**Comment by piaste** [comment]
An alternative approach is to just run a local nuget server - they're very simple and lightweight nowadays.

I use [BaGet](https://loic-sharma.github.io/BaGet/), personally. If you only need a temporary instance for testing, and you have Docker installed, this command suffices:

`docker run -d --name nuget-server -p 5555:80 loicsharma/baget`

Then after each rebuild you run `dotnet nuget push your-nuget-package.nupkg -s http://localhost:5555/v3/index.json` and now you can just add `nuget http://localhost:5555/v3/index.json` to your `paket.dependencies`.

**Comment by NilkOne** [comment]
Hello,
Any solution or worflow to work (dev and debug) with multi Git repo and nuget shared libraries in a super project that depends on shared libs?

**Comment by jakob-ledermann** [comment]
I might have a suggestion:

```
nuget Argu -> git file:///c:\github\Argu feature_branch build:"build.cmd NuGet", Packages: /bin/
```

Now I replaced all occurences of Argu in my `paket.references` files in the project with `# Argu` effectively removing them.
And then I added 
```
<ItemGroup>
  <Reference Include="Argu">
    <HintPath>c:\github\Argu\bin\Debug\Argu.dll</HintPath>
  </Reference>
</ItemGroup>
```
At the end of my `.csproj` files.
Now I can use the new API I introduced.

Could paket not do something simliar automatically?
For me it looks like `Paket.Restore.targets` emits `<PackageReference>`s and then relies on the sdk to resolve them.
So my idea would be some syntax in the paket.local file that causes:
- `<PackageReference>` is not generated for this nuget reference
- instead `<Reference>` or even `<ProjectReference>` tags are generated.

Unfortunately I do not know F# so am in no position to try to get something like this working.
